sklearn/metrics/scorer.py

```diff
@@ -44,7 +45,69 @@
 from ..base import is_regressor
 
 
-class _BaseScorer(metaclass=ABCMeta):
+def _method_caller(cache, estimator, method, *args, **kwargs):
+    """Call estimator with method and args and kwargs."""
+    if cache is None:
+        return getattr(estimator, method)(*args, **kwargs)
+
+    try:
+        return cache[method]
+    except KeyError:
+        result = getattr(estimator, method)(*args, **kwargs)
+        cache[method] = result
+        return result
+
+
+class _MultimetricScorer(dict):
+    """Callable dictionary for multimetric scoring used to avoid repeated calls
+    to `predict_proba`, `predict`, and `decision_function`.
+
+    `_MultimetricScorer` will return a dictionary of scores corresponding to
+    the scorers in the dictionary. Note `_MultimetricScorer` can be created
+    with a dictionary with one key.
+    """
+    def __call__(self, estimator, *args, **kwargs):
+        """Evaluate predicted target values."""
+        scores = {}
+        cache = {} if self._use_cache(estimator) else None
+        method_caller = partial(_method_caller, cache)
```

Comment:
This makes me very sad. Why doesn't have Python a built-in ``apply`` any more?
This is so weird.... having ``_method_caller`` implement ``apply`` seems strange.